### ISSUE TITLE
[MNG-5510] Maven causing ClassCastException with container plugins when project is using other SLF4J implementation than slf4j-simple

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#6821]
**[Juan Pablo Santos Rodríguez](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=juanpablo)** opened **[MNG-5510](https://issues.apache.org/jira/browse/MNG-5510?redirect=false)** and commented

On our project we're using directly logback to start up a custom mbean to be able to change log levels dynamically, among other things. Most of these operations are Logback-specific, they're not available via SLF4J API

For testing we use Jetty, so when we're doing an mvn jetty:run, the mbean gets init'd and we get an Exception stating "ClassCastException: org.slf4j.impl.SimpleLoggerFactory cannot be cast to ch.qos.logback.classic.LoggerContext", due to slf4j-simple being on maven's lib. This is very likely to happen with other container plugins (tomcat7 et al)

---

**Affects:** 3.1.0

**Attachments:**
- [MNG-5510-sample.zip](https://issues.apache.org/jira/secure/attachment/12713400/MNG-5510-sample.zip) (_3.15 kB_)

**Issue Links:**
- [MNG-5393](https://issues.apache.org/jira/browse/MNG-5393) Look at Sonar's use of SLF4J and Logback

2 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

If you have hardcoded requirements for specific SLF4J implementations then I would say that's not great, but to work around to have something that is specific for your use then fork the VM and there will be no conflict.

**Comment by jira-importer** [comment]
**[Juan Pablo Santos Rodríguez](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=juanpablo)** commented

Hi Jason,

let me reopen the issue, as it seems to me it has been misunderstood. The issue is triggered because of having speficic requirements coded against specific SLF4J implementations (btw, f.ex., if you try to reload the log's config file you'll have to go against your specific SLF4J implementation), but it's only an example, as the issue has a broader scope.

We've already made the appropiate changes to bypass the affected code if not on Logback and carried on, but it was just an example of Maven leaking its SLF4J implementation. We had problems with the tomcat and jetty maven plugins when using a different SLF4J implementation, because of having two different SLF4J implementations on classpath.

Another example: go to `$MAVEN_HOME/lib` and change the simple-slf4j jars with another SLF4J implementation. Just because you want to colour maven output, or you want your logs on a database or whatever strange reason you may come up to. As noted at the bottom of \[#1], it's fairly trivial.

Then, when doing a `mvn sonar:sonar` on any maven project, you'll get something like:

```
08:26:09.204 [main] ERROR org.apache.maven.cli.MavenCli Failed to execute goal org.codehaus.mojo:sonar-maven-plugin:2.3:sonar (default-cli) on project YOUR_PROJECT_HERE: Error setting Log implementation.  Cause: java.lang.reflect.InvocationTargetException: org/slf4j/Marker -> [Help 1]
08:26:09.204 [main] ERROR org.apache.maven.cli.MavenCli
```

Forking a new VM may not always be an option, and there may be justified cases which enforce the use of an specific slf4j implementation, leading to this issue. I'm attaching a simple project to reproduce the issue.

\[#1] http://maven.apache.org/maven-logging.html

**Comment by jira-importer** [comment]
**[Juan Pablo Santos Rodríguez](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=juanpablo)** commented

sample project with instructions to reproduce the issue

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

Yes, I know what the issue is. If a host project hard codes an implementation there are going to be issues running from Maven. So your option is to fork for now as I'm not looking at fixing this anytime soon.

**Comment by jira-importer** [comment]
**[Shant Stepanian](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=shantstepanian)** commented

Hi, can you please clarify what you mean by "forking the VM" here?

I don't see any fork option in the sonar plugin (http://mojo.codehaus.org/sonar-maven-plugin/sonar-mojo.html), unlike say maven-surefire-plugin

Is there another methodology you are considering here to fork?

**Comment by jira-importer** [comment]
**[Peter Janes](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=peterj)** commented

The SonarQube occurrence of this issue mentioned in the second comment may be resolved in 5.1 (SONAR-5701/SONAR-5700).
